File: androidchallenge1/src/main/java/class28/androidchallenge01/ui/login/ActLoginRegister.java
package class28.androidchallenge01.ui.login;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import class28.androidchallenge01.R;
import class28.androidchallenge01.utils.ActivityContract;
import class28.androidchallenge01.utils.Utils;

public class ActLoginRegister extends AppCompatActivity implements ActivityContract {

    public static String COME_FROM_RESTAURANT = "COME_FROM_RESTAURANT";
    private FragmentManager fragManager = getSupportFragmentManager();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login_register);
    }

    @Override
    protected void onResume() {
        super.onResume();
        boolean comeFromRestaurant;
        comeFromRestaurant = getIntent().getBooleanExtra(COME_FROM_RESTAURANT, false);
        if (comeFromRestaurant) {
            Utils.startFragment(fragManager, new FragRegister(), R.id.frame_login_register_container);
        } else {
            Utils.startFragment(fragManager, new FragLogin(), R.id.frame_login_register_container);
        }
    }

    @Override
    public void startFragment(Fragment fragment) {
        Utils.startFragment(fragManager, fragment, R.id.frame_login_register_container);
    }

    @Override
    public void onBackPressed() {
        Fragment selectedFragment = null;
        for (Fragment fragment : fragManager.getFragments()) {
            if (fragment.isVisible()) {
                selectedFragment = fragment;
                break;
            }
            return;
        }
        if (selectedFragment instanceof FragLogin) {
            finish();
        } else {
            startFragment(new FragLogin());
        }
    }
}

File: androidchallenge1kotlin/src/main/java/class28/androidchallenge01kotlin/ui/login/ActLoginRegister.kt
package class28.androidchallenge01kotlin.ui.login

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import class28.androidchallenge01kotlin.R
import class28.androidchallenge01kotlin.utils.ActivityContract
import class28.androidchallenge01kotlin.utils.startFragment

class ActLoginRegister : AppCompatActivity(), ActivityContract {

    private val fragManager = supportFragmentManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login_register)
    }

    override fun onResume() {
        super.onResume()
        val comeFromRestaurant: Boolean = intent.getBooleanExtra(COME_FROM_RESTAURANT, false)
        if (comeFromRestaurant) {
            startFragment(fragManager, FragRegister(), R.id.frame_login_register_container)
        } else {
            startFragment(fragManager, FragLogin(), R.id.frame_login_register_container)
        }
    }

    override fun startFragment(fragment: Fragment) {
        startFragment(fragManager, fragment, R.id.frame_login_register_container)
    }

    override fun onBackPressed() {
        var selectedFragment: Fragment? = null
        for (fragment in fragManager.fragments) {
            if (fragment.isVisible) {
                selectedFragment = fragment
                break
            }
            return
        }
        if (selectedFragment is FragLogin) {
            finish()
        } else {
            startFragment(FragLogin())
        }
    }

    companion object {
        var COME_FROM_RESTAURANT = "COME_FROM_RESTAURANT"
    }
}